### ISSUE TITLE
Heap out of bounds read in clnp_print()

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#558]
**Heap out of bounds read in clnp_print()**

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r oobr_clnp_print.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/20091302/5acdcd44-a592-11e6-9167-cd1d8c393e6a.png)

ASAN output:

```
==15898==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60400000e0a7 at pc 0x0000005b6f35 bp 0x7ffe847d3940 sp 0x7ffe847d3938
READ of size 1 at 0x60400000e0a7 thread T0
    #0 0x5b6f34 in clnp_print XYZ/tcpdump/./print-isoclns.c:806:33
    #1 0x5a917a in isoclns_print XYZ/tcpdump/./print-isoclns.c:687:8
    #2 0x566ecc in ethertype_print XYZ/tcpdump/./print-ether.c:357:3
    #3 0x5663f8 in ether_print XYZ/tcpdump/./print-ether.c:229:7
    #4 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #5 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #6 0x7ae607 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #7 0x6b7a7c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #8 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #9 0x7f32cd44d82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #10 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

AddressSanitizer can not describe address in more detail (wild memory access suspected).
SUMMARY: AddressSanitizer: heap-buffer-overflow XYZ/tcpdump/./print-isoclns.c:806:33 in clnp_print
Shadow bytes around the buggy address:
  0x0c087fff9bc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9bd0: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
  0x0c087fff9be0: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fd
  0x0c087fff9bf0: fa fa 00 00 00 00 00 02 fa fa 00 00 00 00 00 01
  0x0c087fff9c00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c087fff9c10: fa fa fa fa[fa]fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==15898==ABORTING
```

**Comment by fxlb** [comment]
Thanks for the report.
This issue is now fixed in the security branch (Will be merged in a few weeks).

**Comment by infrastation** [comment]
You are the original reporter of this vulnerability. Commit 909fb30 fixes this and the fix is available in the release 4.9.0. Thank you very much for the report, please send any future security reports to security@tcpdump.org.
